wurst/data structures/LinkedHashMap.wurst

```diff
@@ -0,0 +1,225 @@
+package LinkedHashMap
+import HashMap
+import LinkedList
+import Wurstunit
+
+public class LinkedHashMap<K, V> extends HashMap<K, V>
+    private LinkedList<K>  keys
+    private LHMEntry<K, V> dummy
+
+    construct()
+        keys  = new LinkedList<K>()
+        dummy = new LHMEntry<K, V>(null, null)
+
+
+    /** Saves the given value under the given key. */
+    override function put(K key, V value)
+        if not has(key)
+            keys.add(key)
+
+        super.put(key, value)
+
+
+    /** Get the number of mappings in this LinkedHashMap. */
+    function size() returns int
+        return keys.getSize()
+
+
+    /** Removes the value saved under the given key. */
+    override function remove(K key)
+        super.remove(key)
+        keys. remove(key)
+
+
+    /** Clears the LinkedHashMap of all entries */
+    override function flush()
+        super.flush()
+        if not keys == null
+            destroy keys
+            keys = new LinkedList<K>()
+
+
+    /** Gets an iterator for this LinkedHashMap. */
+    function iterator() returns LHMIterator<K, V>
+        return new LHMIterator(this, keys)
+
+
+    ondestroy
+        destroy keys
+        destroy dummy
+
+
+public class LHMEntry<K, V>
+    K key
+    V value
+
+    construct(K key, V value)
+        this.key   = key
+        this.value = value
+
+
+public class LHMIterator<K, V>
+    LHMEntry<K, V>      dummy
+    LinkedList<K>       keys
+    LLIterator<K>       keysIterator
+    LinkedHashMap<K, V> parentMap
+
+    construct(LinkedHashMap<K, V> parentMap, LinkedList<K> keys)
+        this.dummy        = new LHMEntry<K, V>(null, null)
+        this.keys         = keys
+        this.keysIterator = keys.iterator()
+        this.parentMap    = parentMap
+
+
+    /** Remove the current element from the LinkedHashMap. */
+    function remove()
+        if not dummy.key == null
+            parentMap.remove(dummy.key)
+            keysIterator.remove()
+
+
+    function hasNext() returns bool
+        return keysIterator.hasNext()
+
+
+    function next() returns LHMEntry<K, V>
+        dummy.key   = keysIterator.next()
+        dummy.value = parentMap.get(dummy.key)
+        return dummy
+
+
+    function close()
+        keysIterator.close()
+        destroy this
```

Comment:
Regarding dealing with destroyed objects: agree it's a programming error,
but some non-class types (like unit) can destroy themselves, so although
the test might not describe it at present, we need some way to deal with
that while iterating without throwing error. I think that if the code were
refactored to use list/map of LHMEntry, as discussed above, then this test
can pass.

Regarding generic tuples: yes that would be nice. It would be nice to have
list functions on tuples as well but that's another subject.

I have for example unit4.forEach extension function in one map.

On Aug 5, 2016 13:42, "Peter Zeller" <notifications@github.com> wrote:

> In wurst/data structures/LinkedHashMap.wurst
> <https://github.com/peq/wurstStdlib/pull/9#discussion_r73685807>:
>
> > +            keysIterator.remove()
> > +
> > +
> > +    function hasNext() returns bool
> > +        return keysIterator.hasNext()
> > +
> > +
> > +    function next() returns LHMEntry<K, V>
> > +        dummy.key   = keysIterator.next()
> > +        dummy.value = parentMap.get(dummy.key)
> > +        return dummy
> > +
> > +
> > +    function close()
> > +        keysIterator.close()
> > +        destroy this
>
> The dummy should be destroyed at the end. It would be even nicer to use a
> tuple instead of the dummy element, but that would require generic tuples
> in Wurst :/
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/peq/wurstStdlib/pull/9/files/079cef9f6e8fc6dfd9ced39e9a320fd464a35583#r73685807>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABWm1hHyiZoe-1q3xFshaiCG_FmpkG58ks5qcy_HgaJpZM4Jc8TL>
> .
>